### DISCUSSION TITLE
Heroku Postgres Integration not working: Error validating datasource `db`: The URL for datasource `db` must start with the protocol `postgresql://`

### DISCUSSION THREAD

**Discussion by daviddalmaso** [prisma/prisma#9539] [Q&A]
Hello!

I'm new to prisma and trying to set up a backend environment on Heroku.

My application is using a react native frontend, Apollo GraphQL for the API layer, and a Node.js backend with Prisma as the ORM connected to a postgres database.

When I run things locally, I have a postgres database set up on my local machine with the DATABASE_URL as `postgresql://<user>:<password>@localhost:5432/mydb?schema=public`. The application works as expected locally.

I've tried to deploy to Heroku using this guide: https://www.prisma.io/docs/guides/deployment/deployment-guides/deploying-to-heroku

I'm able to successfully test my backend server with the Apollo GraphQL playground and see successful API queries to my heroku server.

However, once I point my app to the production server instead of my local machine, I receive the following error:

```
error: Error: error: Error validating datasource `db`: The URL for datasource `db` must start with the protocol `postgresql://`.
  -->  schema.prisma:8
   |
 7 |   provider = "postgresql"
 8 |   url      = env("DATABASE_URL")
   |
```

I did notice that the database set up by the Heroku Postgres add-on has `postgres://` as its protocol instead of `postgresql://`. I didn't think that's an issue but maybe it is. Regardless, Heroku doesn't allow you to change that DATABASE_URL.

I was hoping to get some help on this. Here's the generator and datasource in my `schema.prisma` file:

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["selectRelationCount"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}
```

The migrations all work fine when deploying to the heroku server, but it's just when the Frontend app actually tries to make a call to the backend am I seeing the calls not go through.

Please let me know what other information might help here or if this is a react native issue instead of a prisma issue.

**Comment by ryands17** [comment]
@daviddalmaso 👋 

`postgres://` should also work fine as I use that. Is the environment variable available when you run the application on Heroku? If so could you print it out to the logs and see if it's accessible?

**Reply by daviddalmaso** [reply]
Thanks for getting back!

I'm seeing the correct DATABASE_URL set when I log it during start up

```
2021-10-01T15:24:46.516771+00:00 heroku[web.1]: Starting process with command `node src/server.js`
2021-10-01T15:24:47.639928+00:00 app[web.1]: Database: postgres://*************:*************@ec2-52-6-211-59.compute-1.amazonaws.com:5432/dcci8f69uoacdq
2021-10-01T15:24:48.155024+00:00 app[web.1]: Apollo usage reporting starting! See your graph at https://studio.apollographql.com/graph/Clo-Server-Graph/?variant=current
2021-10-01T15:24:48.157217+00:00 app[web.1]: Apollo schema reporting starting! See your graph at https://studio.apollographql.com/graph/Clo-Server-Graph/?variant=current with server info {"bootId":"a312b075-bd18-407d-be8b-6f49cb7444c2","graphVariant":"current","platform":"local","runtimeVersion":"node v14.18.0","executableSchemaId":"e982197adf83a93059bf4b70a4cbc10e689b571b6f8686b952ea6e241f780a24","serverId":"c42c1aff-cdeb-4d39-92c7-b5e872bf1240","libraryVersion":"apollo-server-core@2.21.0"}
2021-10-01T15:24:48.159054+00:00 app[web.1]: 🚀 Server ready at: http://localhost:38685/
```

Does everything else look fine?

I'm able to successfully make queries when using the ApolloGraphQL playground. It's just my frontend application that is receiving the error message above

**Reply by daviddalmaso** [reply]
One other thing I tried was appending `?schema=public` to the Heroku Postgres DATABASE_URL into a new `DATABASE_URL_WITH_SCHEMA` environment variable.

I updated my `schema.prisma` file datasource to look for this new environment variable.

```
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL_WITH_SCHEMA")
}
```

But I'm still seeing that same error when trying to hit my production server and running my react native app locally:

```
POST RUNQUERY - loading: false, data: undefined, error: Error: error: Error validating datasource `db`: The URL for datasource `db` must start with the protocol `postgresql://`.
  -->  schema.prisma:8
   |
 7 |   provider = "postgresql"
 8 |   url      = env("DATABASE_URL")
   |

Validation Error Count: 1
```

Not sure why it's still looking for the `DATABASE_URL` environment variable there though since I updated the code and deployed to Heroku..

When I start up my Heroku server, it's finding this new `DATABASE_URL_WITH_SCHEMA` as the environment variable

```
2021-10-01T17:56:34.507145+00:00 heroku[web.1]: Starting process with command `node src/server.js`
2021-10-01T17:56:36.177150+00:00 app[web.1]: Database: postgres://************:************@ec2-52-6-211-59.compute-1.amazonaws.com:5432/dcci8f69uoacdq?schema=public
2021-10-01T17:56:37.026561+00:00 app[web.1]: Apollo usage reporting starting! See your graph at https://studio.apollographql.com/graph/Clo-Server-Graph/?variant=current
2021-10-01T17:56:37.030306+00:00 app[web.1]: Apollo schema reporting starting! See your graph at https://studio.apollographql.com/graph/Clo-Server-Graph/?variant=current with server info {"bootId":"b7f831e0-6f4f-4126-8b31-ce5c51044b5b","graphVariant":"current","platform":"local","runtimeVersion":"node v14.18.0","executableSchemaId":"e982197adf83a93059bf4b70a4cbc10e689b571b6f8686b952ea6e241f780a24","serverId":"227a75aa-b70a-460d-96df-33cb968639bd","libraryVersion":"apollo-server-core@2.21.0"}
2021-10-01T17:56:37.034441+00:00 app[web.1]: 🚀 Server ready at: http://localhost:47392
```

**Reply by janpio** [reply]
That the error message still include `DATABASE_URL` is kind of a red flag that the Prisma Client being executed there is somehow not what it should been. Are you importing the correct Prisma Client instance? Is that generated based on the current version of the Prisma Schema? What are you executing that throws this error?

**Reply by daviddalmaso** [reply]
Yeah I agree that the `DATABASE_URL` still being there is a red flag, but after changing the datasource url from `url = env("DATABASE_URL")` to `env("DATABASE_URL_WITH_SCHEMA")` in the `schema.prisma` file and running `prisma migrate dev`, it doesn't create a new migration. Should it be?

This is what importing my prisma client looks on my backend

```
import Prisma from '@prisma/client'
const { PrismaClient } = Prisma

const prisma = new PrismaClient()

console.log(`Database: ${process.env.DATABASE_URL_WITH_SCHEMA}`)

export default prisma
```
I am using `@prisma/client@2.22.1`

The error is being thrown when I run a `useLazyQuery` command on my react native application. That code looks like this:

```
const [runQuery, { loading, data: contactData, error }] = useLazyQuery(USERS_BY_PHONE_NUMBERS)
....
runQuery({ variables: { phoneNumbers } })
console.log(`POST RUNQUERY - loading: ${loading}, data: ${contactData}, error: ${error}`)
```


EDIT:

Here's also my `generator client` in the `schema.prisma` file
```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["selectRelationCount"]
}
```

**Reply by janpio** [reply]
> but after changing the datasource url from `url = env("DATABASE_URL")` to `env("DATABASE_URL_WITH_SCHEMA")` in the `schema.prisma` file and running `prisma migrate dev`, it doesn't create a new migration. Should it be?

Migrations happen when you changed the schema (models) in your `schema.prisma` file.
When you change the configuration of a `generator`, you need to `npx prisma generate` so the Prisma Client is updated to look for that new env variable.

> I am using `@prisma/client@2.22.1`

Uh that is pretty old. If possible, you should update to a much newer version.




> The error is being thrown when I run a `useLazyQuery` command on my react native application. That code looks like this:

I unfortunately can not connect this at all to what is triggering Prisma Client queries in this case. I expect this somehow triggers an API call to some server that uses Prisma Client?

**Comment by janpio** [comment]
> I did notice that the database set up by the Heroku Postgres add-on has `postgres://` as its protocol instead of `postgresql://`. I didn't think that's an issue but maybe it is. Regardless, Heroku doesn't allow you to change that DATABASE_URL.

You can confirm that manually by also changing your local connection string to only have `postgres://` - both should work fine.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
